mscore/instrwidget.cpp

```diff
@@ -478,6 +480,139 @@ void InstrumentsWidget::genPartList(Score* cs)
+            PartiturSelectionType selectionType = _partiturSelectionType;
+            partiturList->selectionModel()->clear();
+            
+            QTreeWidgetItem* item;
+            if (selectionType == PARENTS_SELECTED) {
+                  int idx = delta > 0 ? partiturList->topLevelItemCount() - 1 : 0;
+                  for ( ; (item = partiturList->topLevelItem(idx)); idx += -delta) {
+                        if (!itemsToMove.contains(item))
+                              continue;
+                        
+                              bool isExpanded = partiturList->isItemExpanded(item);
+                              int idx = partiturList->indexOfTopLevelItem(item);
+                              int n = partiturList->topLevelItemCount();
+                              // if part not last, move one slot down
+                              if ((idx + delta < n) && (idx + delta >= 0)) {
+                                    QTreeWidgetItem* item1 = partiturList->takeTopLevelItem(idx);
+                                    // Qt looses the QComboBox set into StaffListItem's when they are re-inserted into the tree:
+                                    // get the currently selected staff type of each combo and re-insert
+                                    int numOfStaffListItems = item1->childCount();
+#if (!defined (_MSCVER) && !defined (_MSC_VER))
+                                    int staffIdx[numOfStaffListItems];
+#else
+                                    // MSVC does not support VLA. Replace with std::vector. If profiling determines that the
+                                    //    heap allocation is slow, an optimization might be used.
+                                    std::vector<int> staffIdx(numOfStaffListItems);
+#endif
+                                    int itemIdx;
+                                    for (itemIdx=0; itemIdx < numOfStaffListItems; ++itemIdx)
+                                          staffIdx[itemIdx] = (static_cast<StaffListItem*>(item1->child(itemIdx)))->staffTypeIdx();
+                                    // do not consider hidden ones
+                                    int plusIdx = delta;
+                                    QTreeWidgetItem* nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    while (nextParent && nextParent->isHidden()) {
+                                          plusIdx += delta;
+                                          nextParent = partiturList->topLevelItem(idx + plusIdx);
+                                    }
+                                    partiturList->insertTopLevelItem(idx + plusIdx, item1);
+                                    // after-re-insertion, recreate each combo and set its index
+                                    for (itemIdx=0; itemIdx < numOfStaffListItems; ++itemIdx) {
+                                          StaffListItem* staffItem = static_cast<StaffListItem*>(item1->child(itemIdx));
```

Comment:
`static_cast` should be avoided if possible. Does `toStaffListItem(item1->child(itemIdx)` work instead? If not, you can stick with static_cast.

Reply:
This was in the original code, only moved to a separate method. I'll have a look.

Reply:
It doesn't exist yet, just checked. Agree to put that in a separate PR later on?

Reply:
Those macros apply only for `ScoreElement` descendants used in libmscore, and those assertions are possible at all just because of type identification system implemented in `ScoreElement` and its descendants.

Using `qobject_cast` here can make sense though since `StaffListItem` [is a QObject](https://github.com/musescore/MuseScore/blob/6226f66e5fd888245b06d210e9558ce49b5d0ff4/mscore/instrwidget.h#L58), and using `qobject_cast` allows run-time type checking for `QObject` descendants. But if it is known for sure that those tree items are `StaffListItem`s then this may be redundant or at least not needed to be changed urgently.